### DISCUSSION TITLE
Build fails

### DISCUSSION THREAD

**Discussion by griz1112** [OpenVSLAM-Community/openvslam_ros#43] [Q&A] [answered]
Trying to build the package it fails with this error. 
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
Socket_Publisher_LIB
    linked by target "run_localization" in directory /home/griz/ovslam_ws/src/openvslam_ros/src
    linked by target "run_slam" in directory /home/griz/ovslam_ws/src/openvslam_ros/src
Which library is this referring too? All the prerequisites are installed.

**Comment by ymd-stella** [comment, marked as answer]
@griz1112 Make sure you have Socket Publisher installed. INSTALL_SOCKET_PUBLISHER must be specified when you install OpenVSLAM.

**Comment by griz1112** [comment]
I did install openvslam that way it's working fine. https://youtu.be/eRZcR8nZU1s Trying to get this integrated into this educational robot see nanosaur.ai via docker container. That vid is openvslam running in that container along with the rest of the nodes to run the bot. Just can't get the ros node to build so I can use the bot's camera.

**Reply by ymd-stella** [reply]
Did you read https://github.com/OpenVSLAM-Community/openvslam_ros/blob/ros2/doc/tutorial_socket.md?

**Comment by griz1112** [comment]
Yea I took that Dockerfile and changed the from ubuntu line to  from  nvidia ubuntu instead.  The video is that set-up running. That part works file its just the ros node won't compile. What library is it looking for I can change the CMakeLists file to look for it. Its running on a jetson nano so building it with Ubuntu won't work.  The nvidia image  already has opencv so commented that part out.  Also built it without docker on an intel box with the same results. Tried the noetic and ros2 openvslam_ros repos. It should work. the ros part is just a wrapper so its weird it doesn't work.  Building everything again without docker to see if I can catch the error that way. I'll let you know how that goes. Noticed a request to implement ros timestamps. Had a similar problem with loam livox so might be able to help out with that once I get this going.

**Reply by ymd-stella** [reply]
> Yea I took that Dockerfile and changed the from ubuntu line to from nvidia ubuntu instead.

Are you confusing https://github.com/OpenVSLAM-Community/openvslam_ros/blob/ros2/Dockerfile.socket with https://github.com/OpenVSLAM-Community/openvslam/blob/main/Dockerfile.socket?

**Comment by griz1112** [comment]
I am using the 2nd one. The first one won't build on the nano. The nano is stuck with 18.04 until q1 of next year. So we build ros2 foxy from source. The debian type install won't work. I tried it and it failed so I built openvslam lib and put the ros wrapper code into the volume I use for all the bot code and tried to build it with colcon there.  I'm using our container with ros2 foxy installed instead of ros:galactic as the base. I will make a few changes and see if the first file you mentioned will work then. Even if you try and run foxy or galactic in a container on the nano you will have kernel issues because its built on 20.04.  Image_transport and cv_bridge are already installed in our container. So delete those add libg2o source and it should work. Thanks for the help I'll let you know how it goes.

**Reply by ymd-stella** [reply]
> Even if you try and run foxy or galactic in a container on the nano you will have kernel issues because its built on 20.04.

I've never experienced it. What happens? I'm interested in the details.

**Comment by griz1112** [comment]
That should have read if its based on 20.04. I had 20.04 installed on my Xavier but it was crashing and there were other problems with it. Mostly to do with the ZED camera SDK and wrapper. So I got in touch with Dustin at Nvidia and he told me to use the jetson-containers repo dockerfiles which use their version of Ubuntu instead of pulling the standard Ubuntu. That fixed  the issues with the ZED but not with the ZED-M or the 2i. The M is small enough to use on the nanosaur so that is the target. A friend at Stereolabs is trying to sort that out. The M and 2i  work with the eloquent container but not on the foxy or galactic container. The ZED wrapper works fine on a native install on 20.04 with foxy or in a container built with 20.04 so its suspected its a kernel issue with the SDK running on nvidia's ubuntu 18.04 in a container. It either hangs or crashes the wrapper. The Nvidia kernel is 4.09 and on 20.04 its 4.15.  Tried that other dockerfile builds but still doesn't run.  So I'm having a friend that is a much better developer than I am look at it. I have a feeling its something to do with our container and not your stuff. I'm a hardware/networking guy for the most part.
